bundles/org.openhab.binding.tr064/src/main/java/org/openhab/binding/tr064/internal/Tr064SubHandler.java

```diff
@@ -0,0 +1,273 @@
+import org.openhab.core.types.RefreshType;
+import org.openhab.core.types.State;
+import org.openhab.core.types.UnDefType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link Tr064SubHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Jan N. Klug - Initial contribution
+ */
+@NonNullByDefault
+public class Tr064SubHandler extends BaseThingHandler {
+    public static final Set<ThingTypeUID> SUPPORTED_THING_TYPES = Set.of(THING_TYPE_SUBDEVICE,
+            THING_TYPE_SUBDEVICE_LAN);
+    private static final int RETRY_INTERVAL = 60;
+
+    private final Logger logger = LoggerFactory.getLogger(Tr064SubHandler.class);
+
+    private Tr064SubConfiguration config = new Tr064SubConfiguration();
+
+    private String deviceType = "";
+    private boolean isInitialized = false;
+
+    private final Map<ChannelUID, Tr064ChannelConfig> channels = new HashMap<>();
+    // caching is used to prevent excessive calls to the same action
+    private final ExpiringCacheMap<ChannelUID, State> stateCache = new ExpiringCacheMap<>(2000);
+    private final Set<ChannelUID> linkedChannels = ConcurrentHashMap.newKeySet();
+
+    private @Nullable SOAPConnector soapConnector;
+    private @Nullable ScheduledFuture<?> connectFuture;
+    private @Nullable ScheduledFuture<?> pollFuture;
+
+    Tr064SubHandler(Thing thing) {
+        super(thing);
+    }
+
+    @Override
+    @SuppressWarnings("null")
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        Tr064ChannelConfig channelConfig = channels.get(channelUID);
+        if (channelConfig == null) {
+            logger.trace("Channel {} not supported.", channelUID);
+            return;
+        }
+
+        if (command instanceof RefreshType) {
+            State state = stateCache.putIfAbsentAndGet(channelUID, () -> soapConnector == null ? UnDefType.UNDEF
+                    : soapConnector.getChannelStateFromDevice(channelConfig, channels, stateCache));
+            if (state != null) {
+                updateState(channelUID, state);
+            }
+            return;
+        }
+
+        if (channelConfig.getChannelTypeDescription().getSetAction() == null) {
+            logger.debug("Discarding command {} to {}, read-only channel", command, channelUID);
+            return;
+        }
+        scheduler.execute(() -> {
+            if (soapConnector == null) {
+                logger.warn("Could not send command because connector not available");
+            } else {
+                soapConnector.sendChannelCommandToDevice(channelConfig, command);
+            }
+        });
+    }
+
+    @Override
+    public void initialize() {
+        config = getConfigAs(Tr064SubConfiguration.class);
+        if (!config.isValid()) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                    "One or more mandatory configuration field is empty");
```

Comment:
This can only happen if the user decides to use textual configuration. So he/she'll probably be an advanced user and be able to figure it out. The UI doesn't allow to save an invalid config.  It would require removing the valid-check in the configuration classes and IMO the code is much clearer this way.